api/acme/order.go

```diff
@@ -0,0 +1,308 @@
+		return o, errors.Errorf("order %s has been abandoned", o.ID)
+	case statusValid:
+		break
+		//return o, errors.Errorf("order %s is already valid", o.ID)
+	case statusPending:
+		return o, errors.Errorf("order %s is not ready", o.ID)
+	case statusReady:
+		break
+	default:
+		return o, errors.Errorf("unexpected status %s for order %s", o.Status, o.ID)
+	}
+
+	// Validate identifier names against CSR alternative names //
+	csrNames := make(map[string]int)
+	if csr.Subject.CommonName != "" {
+		csrNames[csr.Subject.CommonName] = 1
+	}
+	for _, n := range csr.DNSNames {
+		csrNames[n] = 1
+	}
+
+	orderNames := make(map[string]int)
+	for _, n := range o.Identifiers {
+		orderNames[n.Value] = 1
+	}
+
+	if !reflect.DeepEqual(csrNames, orderNames) {
+		// TODO Note reason on the Order error object.
+		return o, errors.Errorf("CSR names do not match order identifiers exactly")
+	}
+
+	// Create and store a new certificate.
+	serverCrt, _, err := auth.Sign(csr, provisioner.Options{
+		NotBefore: provisioner.NewTimeDuration(o.NotBefore),
+		NotAfter:  provisioner.NewTimeDuration(o.NotAfter),
+	})
+
+	_newOrder := *o
+	newOrder := &_newOrder
+	newOrder.Certificate = serverCrt.SerialNumber.String()
+	newOrder.Status = statusValid
+	if err := newOrder.save(db, o); err != nil {
+		return o, err
+	}
+	return newOrder, nil
+}
+
+// GetOrder retrieves and unmarshals an ACME Order type from the database.
+func GetOrder(db nosql.DB, id string) (*Order, error) {
+	b, err := db.Get(orderTable, []byte(id))
+	if err != nil {
+		// TODO return a proper API error indicating bad request.
+		return nil, errors.WithStack(err)
+	}
+	var o Order
+	if err := json.Unmarshal(b, &o); err != nil {
+		return nil, err
+	}
+	return &o, nil
+}
+
+// GetOrderIDsByAccount retrieves a list of Order IDs that were created by the
+// account.
+func GetOrderIDsByAccount(db nosql.DB, id string) ([]string, error) {
+	b, err := db.Get(ordersByAccountIDTable, []byte(id))
+	if nosql.IsErrNotFound(err) {
+		return []string{}, nil
+	} else if err != nil {
+		// TODO return a proper API error indicating bad request.
+		return nil, errors.WithStack(err)
+	}
+	var orderIDs []string
+	if err := json.Unmarshal(b, &orderIDs); err != nil {
+		return nil, err
+	}
+	return orderIDs, nil
+}
+
+type acmeOrder struct {
+	Status         string       `json:"status"`
+	Expires        string       `json:"expires,omitempty"`
+	Identifiers    []Identifier `json:"identifiers"`
+	NotBefore      string       `json:"notBefore,omitempty"`
+	NotAfter       string       `json:"notAfter,omitempty"`
+	Error          interface{}  `json:"error,omitempty"`
+	Authorizations []string     `json:"authorizations"`
+	Finalize       string       `json:"finalize"`
+	Certificate    string       `json:"certificate,omitempty"`
+}
+
+// ToACME converts the internal Order type into the public acmeOrder type for
+// presentation in the ACME protocol.
+func (o *Order) ToACME(db nosql.DB, dir *Directory) (interface{}, error) {
```

Comment:
I think the ToACME methods are used to print JSON, we should be able to rename without a lot of effort this to MarshalJSON, but we might need to inject the arguments to the order (an app context). 

Basically if we want to return an Order as json we should be able to do:
```go
o := NewOrder(...)
// populate o somehow 
JSON(w, o)
```

Reply:
They are used to output JSON, but I'm already using the default json marshaler when writing the objects to the DB. The difference here is that a number of fields are hidden when output as an acme response (e.g. ID, Created, etc.)

Reply:
It will be easy to clean this once the logic is split on a different layer, the return object of ToACME should be the return object of the method doing the logic.